exporter/splunkhecexporter/client.go

```diff
@@ -36,25 +37,156 @@ import (
 	"github.com/open-telemetry/opentelemetry-collector-contrib/internal/splunk"
 )
 
+var (
+	errOverCapacity = errors.New("over capacity")
+)
+
+// Minimum number of bytes to compress. 1500 is the MTU of an ethernet frame.
+const minCompressionLen = 1500
+
 // client sends the data to the splunk backend.
 type client struct {
-	config  *Config
-	url     *url.URL
-	client  *http.Client
-	logger  *zap.Logger
-	zippers sync.Pool
-	wg      sync.WaitGroup
-	headers map[string]string
+	config         *Config
+	url            *url.URL
+	client         *http.Client
+	logger         *zap.Logger
+	wg             sync.WaitGroup
+	headers        map[string]string
+	gzipWriterPool *sync.Pool
 }
 
 // bufferState encapsulates intermediate buffer state when pushing data
 type bufferState struct {
-	buf      *bytes.Buffer
-	tmpBuf   *bytes.Buffer
-	bufFront *index
-	bufLen   int
-	resource int
-	library  int
+	compressionAvailable bool
+	compressionEnabled   bool
+	bufferMaxLen         uint
+	writer               io.Writer
+	buf                  *bytes.Buffer
+	bufFront             *index
+	resource             int
+	library              int
+	gzipWriterPool       *sync.Pool
+}
+
+func (b *bufferState) reset() {
+	b.buf.Reset()
+	b.compressionEnabled = false
+	b.writer = &cancellableBytesWriter{innerWriter: b.buf, maxCapacity: b.bufferMaxLen}
+}
+
+func (b *bufferState) Read(p []byte) (n int, err error) {
+	return b.buf.Read(p)
+}
+
+func (b *bufferState) Close() error {
+	if _, ok := b.writer.(*cancellableGzipWriter); ok {
+		return b.writer.(*cancellableGzipWriter).close()
+	}
+	return nil
+}
+
+// accept returns true if data is accepted by the buffer
+func (b *bufferState) accept(data []byte) bool {
+	_, err := b.writer.Write(data)
+	overCapacity := errors.Is(err, errOverCapacity)
+	bufLen := b.buf.Len()
+	if overCapacity {
+		bufLen += len(data)
+	}
```

Comment:
Not sure if it's the right logic here. If we don't have any error or error is not `errOverCapacity`, we don't increase `bufLen`?

Reply:
If the error is present, then `bufLen`, which is set at the previous line to the length of the buffer, is missing the additional bytes from the data, as write failed. We add the length of the bytes we are trying to write to the `bufLen` variable.

Reply:
It's a corner case - here is how it would work out:
We have a `cancellableBytesWriter` that contains 1400 bytes, and we set max content length to 1800. We now try to write additional 500 bytes. The writer refuses and returns an error, `errOverCapacity`. Because it refused, it didn't write the bytes into its buffer. The buffer is still containing 1400 bytes.

When we look at the length of the buffer to determine if it's worth compressing, we compare it to an arbitrary minimum length of 1500 bytes. To make sure we understand the true number of bytes we are looking to compress we must first make sure to add 500 bytes to the known buffer length since they failed to write. 1500 < 1400+500, and therefore we know to use compression.

Reply:
I see. Makes sense. Thanks for explaining